crypto/bn/bn_prime.c

```diff
@@ -478,98 +447,48 @@ int bn_probable_prime_dh(BIGNUM *rnd, int bits,
     if (!BN_sub(rnd, rnd, t1))
         goto err;
     if (rem == NULL) {
-        if (!BN_add_word(rnd, 1))
+        if (!BN_add_word(rnd, safe ? 3u : 1u))
```

Comment:
I don't see why 3 would be a better default in case it's for a safe prime.

Reply:
this kind of assumes that add is 4 or something similar.
The old code made this: q = 1xxxxx1, and then p = 2*q+1, therefore
p had the form 1xxxxx11, so this is just what implicitly happened previously.

Reply:
And the documentation says it's 1. I think this is the sort of behaviour we shouldn't do, and we should just have generated an error in that case.

Reply:
I guess, the documentation was wrong all the time.
I was concerned about what happens if someone did
```
BN_set_word(i4,4);
BN_generate_prime_ex(r,bits,1,i4,NULL,cb);
```
Unless I am mistaken this did return r=3 mod 4,
and would run forever if I make it look for r=1 mod 4
(unless maybe for bits=3 because of 5 being a safe prime)

Reply:
this is the current code (probable_prime_dh_safe)
```C
    if (rem == NULL) {
        if (!BN_add_word(q, 1))
            goto err;
    } else {
    ...
    }

    /* we now have a random number 'rand' to test. */
    if (!BN_lshift1(p, q))
        goto err;
    if (!BN_add_word(p, 1))
        goto err;
```

so if rem==NULL we have  p= 2*(q+1)+1 = 2*q + 3 and q = q+1
I guess I should update the documentation.